datafusion/physical-plan/src/projection.rs

```diff
@@ -306,9 +306,10 @@ impl ProjectionStream {
         let arrays = self
             .expr
             .iter()
-            .map(|expr| {
+            .zip(&self.schema.fields)
+            .map(|(expr, field)| {
                 expr.evaluate(batch)
-                    .and_then(|v| v.into_array(batch.num_rows()))
+                    .and_then(|v| v.into_array_of_type(batch.num_rows(), field.data_type()))
```

Comment:
Not sure if casting here is a correct approach.

Reply:
It looks incorrect to me 🤔